// Market Regime Service - Classifies current market conditions using Claude
// Part of BLOASIS Tier 1 (shared across all users, cached for 6 hours)

syntax = "proto3";

package bloasis.market_regime;

import "google/api/annotations.proto";
import "common.proto";

// MarketRegimeService provides market regime classification
// Regime classification is shared across ALL users (Tier 1) and cached for cost optimization
service MarketRegimeService {
  // GetCurrentRegime returns the current market regime classification
  // Uses Claude for analysis, cached for 6 hours unless force_refresh is true
  rpc GetCurrentRegime(GetCurrentRegimeRequest) returns (GetCurrentRegimeResponse) {
    option (google.api.http) = {
      get: "/v1/market-regime/current"
    };
  }

  // GetRegimeHistory returns historical regime classifications within a time range
  // Useful for backtesting and analyzing regime transitions
  rpc GetRegimeHistory(GetRegimeHistoryRequest) returns (GetRegimeHistoryResponse) {
    option (google.api.http) = {
      get: "/v1/market-regime/history"
    };
  }

  // SaveRegime saves a new regime classification to the database
  // Internal use only: Called by schedulers or other backend services
  // No HTTP annotation - not exposed via REST
  rpc SaveRegime(SaveRegimeRequest) returns (SaveRegimeResponse);
}

// GetCurrentRegimeRequest - Request for current market regime
message GetCurrentRegimeRequest {
  // If true, bypass cache and force fresh analysis from Claude
  // Use sparingly to avoid unnecessary API costs
  bool force_refresh = 1;
}

// Market indicators used in regime classification
message MarketIndicators {
  // VIX volatility index
  double vix = 1;

  // S&P 500 trend direction
  string sp500_trend = 2;

  // Yield curve spread (10Y - 2Y)
  string yield_curve = 3;

  // Credit spreads indicator
  string credit_spreads = 4;
}

// GetCurrentRegimeResponse - Current market regime classification
message GetCurrentRegimeResponse {
  // Market regime classification:
  // - "crisis": High volatility, risk-off (VIX > 30, major drawdowns)
  // - "bear": Declining market, moderate volatility
  // - "bull": Rising market, low-moderate volatility
  // - "sideways": Range-bound, no clear direction
  // - "recovery": Transition from crisis/bear to bull
  string regime = 1;

  // Confidence score from AI classification (0.0 to 1.0)
  // Higher values indicate stronger signal clarity
  double confidence = 2;

  // ISO 8601 timestamp of when this regime was classified
  // Example: "2025-01-26T14:30:00Z"
  string timestamp = 3;

  // What triggered this regime classification:
  // - "baseline": Scheduled analysis (every 6 hours)
  // - "fomc": Federal Reserve meeting/announcement
  // - "circuit_breaker": Market halt or extreme volatility spike
  // - "earnings_season": Major earnings period affecting sentiment
  // - "geopolitical": Major geopolitical event
  string trigger = 4;

  // AI-generated reasoning explaining the regime classification
  string reasoning = 5;

  // Risk level assessment: "low", "medium", or "high"
  string risk_level = 6;

  // Market indicators used in this classification
  MarketIndicators indicators = 7;
}

// GetRegimeHistoryRequest - Request for historical regime data
message GetRegimeHistoryRequest {
  // Time range for historical query
  // Uses bloasis.common.TimeRange for ISO 8601 date format
  bloasis.common.TimeRange time_range = 1;
}

// GetRegimeHistoryResponse - Historical regime classifications
message GetRegimeHistoryResponse {
  // List of regime classifications within the requested time range
  // Ordered chronologically (oldest first)
  repeated GetCurrentRegimeResponse regimes = 1;
}

// SaveRegimeRequest - Request to save a regime classification
message SaveRegimeRequest {
  // Market regime classification (crisis, bear, bull, sideways, recovery)
  string regime = 1;

  // Confidence score (0.0 to 1.0)
  double confidence = 2;

  // What triggered this classification (baseline, fomc, circuit_breaker, etc.)
  string trigger = 3;

  // Optional: ISO 8601 timestamp. If not provided, current time is used.
  string timestamp = 4;
}

// SaveRegimeResponse - Response after saving regime
message SaveRegimeResponse {
  // Whether the operation was successful
  bool success = 1;

  // The saved regime data
  GetCurrentRegimeResponse regime = 2;
}
